File: src/main/services/optionsService.ts
import { existsSync, readFileSync, writeFileSync } from 'fs'
import { join } from 'path'

// Key-value pairs to force regardless of what options.txt currently says
const PERF_OVERRIDES: Record<string, string> = {
  maxFps:                  '260',     // 260 = unlimited in MC 1.17+
  enableVsync:             'false',
  renderDistance:          '8',
  simulationDistance:      '6',
  particles:               '2',       // minimal
  cloudRenderMode:         'off',
  entityShadows:           'false',
  ao:                      'false',
  biomeBlendRadius:        '0',
}

export function patchOptionsFile(gameDir: string): void {
  const optPath = join(gameDir, 'options.txt')

  let lines: string[] = []
  if (existsSync(optPath)) {
    lines = readFileSync(optPath, 'utf-8').split('\n')
  }

  // Build a map of existing entries
  const map = new Map<string, string>()
  for (const line of lines) {
    const colon = line.indexOf(':')
    if (colon === -1) continue
    map.set(line.slice(0, colon).trim(), line.slice(colon + 1).trim())
  }

  // Apply overrides
  for (const [key, value] of Object.entries(PERF_OVERRIDES)) {
    map.set(key, value)
  }

  // Reconstruct file
  const out = [...map.entries()].map(([k, v]) => `${k}:${v}`).join('\n') + '\n'
  writeFileSync(optPath, out, 'utf-8')
}

File: src-tauri/src/services/options_service.rs
//! Port of `src/main/services/optionsService.ts`.

use std::collections::BTreeMap;
use std::fs;
use std::path::Path;

/// Key-value pairs forced regardless of what options.txt currently says. `BTreeMap` (vs. the
/// TS `Record`) keeps insertion-independent, deterministic output — not load-bearing, just tidy.
fn perf_overrides() -> BTreeMap<&'static str, &'static str> {
    BTreeMap::from([
        ("maxFps", "260"), // 260 = unlimited in MC 1.17+
        ("enableVsync", "false"),
        ("renderDistance", "8"),
        ("simulationDistance", "6"),
        ("particles", "2"), // minimal
        ("cloudRenderMode", "off"),
        ("entityShadows", "false"),
        ("ao", "false"),
        ("biomeBlendRadius", "0"),
    ])
}

pub fn patch_options_file(game_dir: &Path) {
    let opt_path = game_dir.join("options.txt");
    let existing = fs::read_to_string(&opt_path).unwrap_or_default();

    // Preserve insertion order of the existing file, only appending new keys at the end —
    // matches the TS `Map` behaviour (Map preserves first-insertion order on re-`set`).
    let mut order: Vec<String> = Vec::new();
    let mut map: BTreeMap<String, String> = BTreeMap::new();
    let mut ordered_map: Vec<(String, String)> = Vec::new();

    for line in existing.lines() {
        if let Some(colon) = line.find(':') {
            let key = line[..colon].trim().to_string();
            let value = line[colon + 1..].trim().to_string();
            if !map.contains_key(&key) {
                order.push(key.clone());
            }
            map.insert(key, value);
        }
    }

    for (key, value) in perf_overrides() {
        if !map.contains_key(key) {
            order.push(key.to_string());
        }
        map.insert(key.to_string(), value.to_string());
    }

    for key in &order {
        if let Some(value) = map.get(key) {
            ordered_map.push((key.clone(), value.clone()));
        }
    }

    let out = ordered_map.iter().map(|(k, v)| format!("{k}:{v}")).collect::<Vec<_>>().join("\n") + "\n";
    let _ = fs::write(&opt_path, out);
}
